net.c

```diff
@@ -7,143 +7,237 @@
     }
 
-    for(ai = airoot; ai; ai = ai->ai_next) {
-      fd = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
-      if (fd == -1) {
-        twarn("socket()");
-        continue;
-      }
+    for (ai = airoot; ai; ai = ai->ai_next) {
+        fd = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
+        if (fd == -1) {
+            twarn("socket()");
+            continue;
+        }
 
-      flags = fcntl(fd, F_GETFL, 0);
-      if (flags < 0) {
-        twarn("getting flags");
-        close(fd);
-        continue;
-      }
+        r = set_nonblocking(fd);
+        if (r == -1) {
+            close(fd);
+            continue;
+        }
 
-      r = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
-      if (r == -1) {
-        twarn("setting O_NONBLOCK");
-        close(fd);
-        continue;
-      }
+        flags = 1;
+        r = setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &flags, sizeof flags);
+        if (r == -1) {
+            twarn("setting SO_REUSEADDR on fd %d", fd);
+            close(fd);
+            continue;
+        }
+        r = setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, &flags, sizeof flags);
+        if (r == -1) {
+            twarn("setting SO_KEEPALIVE on fd %d", fd);
+            close(fd);
+            continue;
+        }
+        r = setsockopt(fd, SOL_SOCKET, SO_LINGER, &linger, sizeof linger);
+        if (r == -1) {
+            twarn("setting SO_LINGER on fd %d", fd);
+            close(fd);
+            continue;
+        }
+        r = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &flags, sizeof flags);
+        if (r == -1) {
+            twarn("setting TCP_NODELAY on fd %d", fd);
+            close(fd);
+            continue;
+        }
 
-      flags = 1;
-      r = setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &flags, sizeof flags);
-      if (r == -1) {
-        twarn("setting SO_REUSEADDR on fd %d", fd);
-        close(fd);
-        continue;
-      }
-      r = setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, &flags, sizeof flags);
-      if (r == -1) {
-        twarn("setting SO_KEEPALIVE on fd %d", fd);
-        close(fd);
-        continue;
-      }
-      r = setsockopt(fd, SOL_SOCKET, SO_LINGER, &linger, sizeof linger);
-      if (r == -1) {
-        twarn("setting SO_LINGER on fd %d", fd);
-        close(fd);
-        continue;
-      }
-      r = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &flags, sizeof flags);
-      if (r == -1) {
-        twarn("setting TCP_NODELAY on fd %d", fd);
+        r = bind(fd, ai->ai_addr, ai->ai_addrlen);
+        if (r == -1) {
+            twarn("bind()");
+            close(fd);
+            continue;
+        }
+        if (verbose) {
+            char hbuf[NI_MAXHOST], pbuf[NI_MAXSERV], *h = host, *p = port;
+            struct sockaddr_in addr;
+            socklen_t addrlen;
+
+            addrlen = sizeof(addr);
+            r = getsockname(fd, (struct sockaddr *) &addr, &addrlen);
+            if (!r) {
+                r = getnameinfo(ai->ai_addr, ai->ai_addrlen,
```

Comment:
Original line is `             r = getnameinfo((struct sockaddr *) &addr, addrlen,`, your version looks too complicated. What is purpose of writing into `ai->ai_addr` instead of local var `addr`?

Reply:
Again, an accident.  I never had the intention to modify the original code for the listening socket setup.  I trusted the automatic rebase and Emacs three way merge more than I should have.